staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -29,30 +32,116 @@ import (
+	identifier runtime.Identifier
+}
+
+func newWatchEmbeddedEncoder(ctx context.Context, encoder runtime.Encoder, target *schema.GroupVersionKind, options interface{}, scope *RequestScope) *watchEmbeddedEncoder {
+	result := &watchEmbeddedEncoder{
+		encoder: encoder,
+		ctx:     ctx,
+		target:  target,
+	}
+	if target != nil && target.Kind == "Table" {
+		if tableOptions, ok := options.(*metav1.TableOptions); ok && tableOptions != nil {
+			result.options = tableOptions
+		}
+		result.scope = scope
+	}
+	return result
+}
+
+// Encode implements runtime.Encoder interface.
+func (e *watchEmbeddedEncoder) Encode(obj runtime.Object, w io.Writer) error {
 	if co, ok := obj.(runtime.CacheableObject); ok {
-		if target != nil {
-			// Non-nil mediaType.Convert means that some conversion of the object
-			// has to happen. Currently conversion may potentially modify the
-			// object or assume something about it (e.g. asTable operates on
-			// reflection, which won't work for any wrapper).
-			// To ensure it will work correctly, let's operate on base objects
-			// and not cache it for now.
-			//
-			// TODO: Long-term, transformObject should be changed so that it
-			// implements runtime.Encoder interface.
-			return doTransformObject(ctx, co.GetObject(), opts, target, scope)
+		return co.CacheEncode(e.Identifier(), e.doEncode, w)
+	}
+	return e.doEncode(obj, w)
+}
+
+func (e *watchEmbeddedEncoder) doEncode(obj runtime.Object, w io.Writer) error {
+	result, err := doTransformObject(e.ctx, obj, e.options, e.target, e.scope)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to transform object %v: %v", reflect.TypeOf(obj), err))
+		result = obj
+	}
+
+	// When we are tranforming to a table, use the original table options when
+	// we should print headers only on the first object - headers should be
+	// omitted on subsequent events.
+	if tableOptions, ok := e.options.(*metav1.TableOptions); ok && !tableOptions.NoHeaders {
+		tableOptions.NoHeaders = true
+		// With options change, we should recompute the identifier.
+		// Clearing this will trigger lazy recompute when needed.
+		e.identifier = ""
+	}
+
+	return e.encoder.Encode(result, w)
+}
+
+// Identifier implements runtime.Encoder interface.
+func (e *watchEmbeddedEncoder) Identifier() runtime.Identifier {
+	if e.identifier == "" {
+		e.identifier = e.embeddedIdentifier()
+	}
+	return e.identifier
+}
+
+type watchEmbeddedEncoderIdentifier struct {
+	Name    string              `json:"name,omitempty"`
+	Encoder string              `json:"encoder,omitempty"`
+	Target  string              `json:"target,omitempty"`
+	Options metav1.TableOptions `json:"options,omitempty"`
+	Scope   string              `json:"scope,omitempty"`
+}
+
+func (e *watchEmbeddedEncoder) embeddedIdentifier() runtime.Identifier {
+	if e.target == nil {
+		// If no conversion is performed, we effective only use
+		// the embedded identifier.
+		return e.encoder.Identifier()
```

Comment:
add a comment to the target == nil case in `doTransformObject` that if we ever change that from a no-op, this needs to update to include whatever variable behavior we have

Reply:
done